File: java/atbash-cipher/src/main/java/Atbash.java
import java.util.function.IntUnaryOperator;

public class Atbash {
    public String encode(String input) {
        return groupBy5(__code(input));
    }

    public String decode(String input) {
        return __code(input);
    }

    // convert 'a' to 'z', 'b' to 'y', etc
    private static IntUnaryOperator encoder = c ->
            Character.isLetter(c)
                    ? 'a' + (26 - (c - ('a' - 1)))
                    : c;

    private String __code(String input) {
        return input
                .codePoints()
                .filter(Character::isLetterOrDigit)
                .map(Character::toLowerCase)
                .map(encoder)
                .collect(
                        StringBuilder::new,
                        StringBuilder::appendCodePoint,
                        StringBuilder::append
                )
                .toString();
    }

    private String groupBy5(String input) {
        return input
                .replaceAll(".{1,5}", "$0 ")
                .replaceFirst(" $", "");
    }
}

File: kotlin/atbash-cipher/src/main/kotlin/Atbash.kt
object Atbash {

    fun encode(s: String): String = decode(s).groupsOf()

    fun decode(s: String): String =
            s.filter(Character::isLetterOrDigit)    // s.replace(Regex("[^\\p{Alnum}]"), "")
                    .toLowerCase()
                    .map { encoding[it] ?: it }
                    .joinToString("")

    // I now realize I'm reimplementing String::chunked
    private fun String.groupsOf(size: Int = 5): String =
        Regex(".{1,$size}")
                .findAll(this)
                .map(MatchResult::value)
                .joinToString(" ")

    private val encoding = with('a'..'z') { this.zip(this.reversed()).toMap() }
}
